src/github.com/getlantern/balancer/dialer.go

```diff
@@ -94,44 +78,94 @@ func (d *dialer) start() {
 					d.OnClose()
 				}
 				return
-			case t := <-d.errCh:
-				lastFailed = t
-			case <-timer.C:
-				ok := d.Check()
-				if ok {
-					lastCheckSucceeded = time.Now()
-					timer.Reset(longDuration)
-					consecCheckFailures = 0
-				} else {
-					consecCheckFailures += 1
-				}
+			case <-d.errCh:
+				d.markFailure()
+			case <-d.checkTimer.C:
+				go func() {
```

Comment:
The purpose is to calculate metrics even if the dialer is not picked recently by the strategy. Check in same goroutine will block the loop, maybe spawn a dedicated goroutine? Will try.